bayesian-binning.1
'\" t
.\"     Title: BAYESIAN_BINNING
.\"    Author: Philipp Benner <philipp.benner@mis.mpg.de>
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/07/2012
.\"    Manual: Bayesian Binning User Manual
.\"    Source: bayesian-binning
.\"  Language: English
.\"
.TH "BAYESIAN_BINNING" "1" "01/07/2012" "bayesian-binning" "Bayesian Binning User Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
bayesian-binning \- Bayesian binning of well\-ordered data
.SH "SYNOPSIS"
.HP \w'\fBbayesian\-binning\fR\ 'u
\fBbayesian\-binning\fR \fB\-e\ \fR\fB\fIthis\fR\fR [\fB\-\-example=\fR\fBthat\fR] [{\fB\-e\fR\ |\ \fB\-\-example\fR}\ \fIthis\fR] [{\fB\-e\fR\ |\ \fB\-\-example\fR}\ {\fIthis\fR\ |\ \fIthat\fR}]
.HP \w'\fBbayesian\-binning\fR\ 'u
\fBbayesian\-binning\fR [{\fB\-h\fR\ |\ \fB\-\-help\fR} | {\fB\-v\fR\ |\ \fB\-\-version\fR}]
.SH "DESCRIPTION"
.PP
This manual page documents briefly the
\fBbayesian\-binning\fR
and
\fBbar\fR
commands\&.
.PP
This manual page was written for the Debian distribution because the original program does not have a manual page\&. Instead, it has documentation in the GNU
\fBinfo\fR(1)
format; see below\&.
.PP
\fBbayesian\-binning\fR
is a program that\&.\&.\&.
.SH "OPTIONS"
.PP
The program follows the usual GNU command line syntax, with long options starting with two dashes (`\-\*(Aq)\&. A summary of options is included below\&. For a complete description, see the
\fBinfo\fR(1)
files\&.
.PP
\fB\-e \fR\fB\fIthis\fR\fR, \fB\-\-example=\fR\fB\fIthat\fR\fR
.RS 4
Does this and that\&.
.RE
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Show summary of options\&.
.RE
.PP
\fB\-v\fR, \fB\-\-version\fR
.RS 4
Show version of program\&.
.RE
.SH "FILES"
.PP
/etc/foo\&.conf
.RS 4
The system\-wide configuration file to control the behaviour of
bayesian\-binning\&. See
\fBfoo.conf\fR(5)
for further details\&.
.RE
.PP
${HOME}/\&.foo\&.conf
.RS 4
The per\-user configuration file to control the behaviour of
bayesian\-binning\&. See
\fBfoo.conf\fR(5)
for further details\&.
.RE
.SH "ENVIONMENT"
.PP
\fBFOO_CONF\fR
.RS 4
If used, the defined file is used as configuration file (see also
the section called \(lqFILES\(rq)\&.
.RE
.SH "DIAGNOSTICS"
.PP
The following diagnostics may be issued on
stderr:
.PP
Bad configuration file\&. Exiting\&.
.RS 4
The configuration file seems to contain a broken configuration line\&. Use the
\fB\-\-verbose\fR
option, to get more info\&.
.RE
.PP
\fBbayesian\-binning\fR
provides some return codes, that can be used in scripts:
.\" line length increase to cope w/ tbl weirdness
.ll +(\n(LLu * 62u / 100u)
.TS
ll.
\fICode\fR	\fIDiagnostic\fR
T{
\fB0\fR
T}	T{
Program exited successfully\&.
T}
T{
\fB1\fR
T}	T{
The configuration file seems to be broken\&.
T}
.TE
.\" line length decrease back to previous value
.ll -(\n(LLu * 62u / 100u)
.sp
.SH "BUGS"
.PP
The program is currently limited to only work with the
foobar
library\&.
.PP
The upstreams
BTS
can be found at
\m[blue]\fB\%http://bugzilla.foo.tld\fR\m[]\&.
.SH "SEE ALSO"
.PP
\fBbar\fR(1),
\fBbaz\fR(1),
\fBfoo.conf\fR(5)
.PP
The programs are documented fully by
The Rise and Fall of a Fooish Bar
available via the
\fBinfo\fR(1)
system\&.
.SH "AUTHOR"
.PP
\fBPhilipp Benner\fR <\&philipp\&.benner@mis\&.mpg\&.de\&>
.RS 4
Wrote this manpage for the Debian system\&.
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2007 Philipp Benner
.br
.PP
This manual page was written for the Debian system (but may be used by others)\&.
.PP
Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 2 or (at your option) any later version published by the Free Software Foundation\&.
.PP
On Debian systems, the complete text of the GNU General Public License can be found in
/usr/share/common\-licenses/GPL\&.
.sp


ChangeLog
updatedd (2.6) unstable; urgency=low

  * getopt_long implementation

 -- Philipp Benner <mail@philipp-benner.de>  Sun,  3 Sep 2006 22:45:55 +0200

updatedd (2.5) unstable; urgency=low

  * makefiles cross-compile compatible
  * implemented gnu makefile conventions

 -- Philipp Benner <mail@philipp-benner.de>  Thu, 12 May 2005 21:22:32 +0200

updatedd (2.4) unstable; urgency=low

  * create $(bindir) when it doesn't exist
  * use $(DESTDIR) when installing updatedd
  * changing the cache file of upatedd-wrapper didn't work
  * ported the source to Darwin

 -- Philipp Benner <mail@philipp-benner.de>  Tue, 08 Feb 2005 20:07:25 +0100

updatedd (2.3) unstable; urgency=low

  * check for error.h

 -- Philipp Benner <mail@philipp-benner.de>  Sun, 10 Oct 2004 15:58:55 +0100

updatedd (2.2-1) unstable; urgency=low

  * buggy Makefile fixed
  * corrected some mistakes in the documentation

 -- Philipp Benner <mail@philipp-benner.de>  Wed, 11 Aug 2004 18:06:27 +0100

updatedd (2.1-1) unstable; urgency=low

  * minor bugfixes

 -- Philipp Benner <mail@philipp-benner.de>  Sat, 01 May 2004 13:40:45 +0100

updatedd (2.0-1) unstable; urgency=medium

  * coded updatedd from scratch
  * new application: updatedd-wrapper
  * config file support
  * new cache file support
  * support for new services
  * support for various check-ip servers

 -- Philipp Benner <mail@philipp-benner.de>  Fri, 14 Feb 2004 15:13:30 +0100

updatedd (1.9-2) unstable; urgency=low

  * dyndns.org update

 -- Philipp Benner <mail@philipp-benner.de>  Fri, 13 Feb 2004 15:13:30 +0100

updatedd (1.9-1) unstable; urgency=low

  * support for regfish.com
  * new return codes
  * new scripts
  * french documentation

 -- Philipp Benner <mail@philipp-benner.de>  Sun, 08 Feb 2004 10:35:01 +0100

updatedd (1.9pre1-1) unstable; urgency=low

  * removed loop which caused extra traffic
  * bug fix in noip.c

 -- Philipp Benner <mail@philipp-benner.de>  Tue, 22 Oct 2003 09:41:25 +0100

updatedd (1.8-1) unstable; urgency=low

  * dyndns checkip bug fixed
  * new code structure
  * implementation of getopt_long() for *BSD
  * new rc scripts

 -- Philipp Benner <mail@philipp-benner.de>  Tue, 14 Oct 2003 09:41:25 +0100

updatedd (1.7-1) unstable; urgency=low

  * dyndns checkip bug fixed

 -- Philipp Benner <mail@philipp-benner.de>  Fri, 3 Oct 2003 23:22:32 +0100


bayesian-binning.py
#! /usr/bin/env python

# Copyright (C) 2010 Philipp Benner
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import sys
import getopt
import os
import ConfigParser
import numpy as np
import math
from itertools import izip
from matplotlib import *
from matplotlib.pyplot import *
import matplotlib.patches as patches
import matplotlib.path as path

import bayesian_binning.config as config
import bayesian_binning.interface as interface
import bayesian_binning.visualization as vis
import bayesian_binning.statistics as statistics

# global options
# ------------------------------------------------------------------------------

verbose = False

# usage
# ------------------------------------------------------------------------------

def usage():
    """Print usage."""
    print
    print "bayesian-binning.py [option]... FILE "
    print
    print "Options:"
    print "   -b                             - compute break probabilities"
    print "   -d                             - compute differential gain"
    print "   -e                             - compute multibin entropies"
    print "       --effective-counts         - compute effective counts"
    print "   -m  --marginal                 - compute full marginal distribution"
    print "   -r  --marginal-range=(FROM,TO) - limit range for the marginal distribution"
    print "   -s  --marginal-step=STEP       - step size for the marginal distribution"
    print "       --epsilon=EPSILON          - epsilon for entropy estimations"
    print "   -k  --moments=N                - compute the first N>=3 moments"
    print "       --which=EVENT              - for which event to compute the binning"
    print
    print "       --load=FILE                - load result from file"
    print "       --save=FILE                - save result to file"
    print
    print "   -h, --help                     - print help"
    print "   -v, --verbose                  - be verbose"
    print "   -t, --prombsTest               - test prombs algorithm"
    print

# load results from file
# ------------------------------------------------------------------------------

def load_config():
    config_parser = ConfigParser.RawConfigParser()
    config_parser.read(options['load'])
    if not config_parser.has_section('Result'):
        raise IOError("Invalid configuration file.")

    moments   = config.readMatrix(config_parser, 'Result', 'moments',   float)
    marginals = config.readMatrix(config_parser, 'Result', 'marginals', float)
    mpost     = config.readVector(config_parser, 'Result', 'mpost',     float)
    if config_parser.has_option('Result', 'bprob'):
        bprob = config.readVector(config_parser, 'Result', 'bprob',     float)
    else:
        bprob     = []
    if config_parser.has_option('Result', 'multibin_entropy'):
        multibin_entropy = config.readVector(config_parser, 'Result', 'multibin_entropy', float)
    else:
        multibin_entropy = []
    result = {
        'moments'   : moments,
        'marginals' : marginals,
        'bprob'     : bprob,
        'mpost'     : mpost,
        'multibin_entropy'  : multibin_entropy,
        'differential_gain' : [] }
    return result

# binning
# ------------------------------------------------------------------------------

def bin(counts, alpha, mprior):
    """Call the binning library."""
    if options['load']:
        return load_config()
    else:
        counts_i = [ map(int, row) for row in counts ]
        alpha_i  = [ map(int, row) for row in alpha  ]
        mprior_i =   map(float, mprior)
        return interface.binning(counts_i, alpha_i, mprior_i, options)

# save result
# ------------------------------------------------------------------------------

def saveResult(result):
    config = ConfigParser.ConfigParser()
    config.add_section('Result')
    config.set('Result', 'moments', "\n"+"\n".join(map(lambda arg: " ".join(map(str, arg)), result['moments'])))
    config.set('Result', 'marginals', "\n"+"\n".join(map(lambda arg: " ".join(map(str, arg)), result['marginals'])))
    config.set('Result', 'bprob',   " ".join(map(str, result['bprob'])))
    config.set('Result', 'mpost',   " ".join(map(str, result['mpost'])))
    config.set('Result', 'multibin_entropy', " ".join(map(str, result['multibin_entropy'])))
    configfile = open(options['save'], 'wb')
    config.write(configfile)

# parse config file
# ------------------------------------------------------------------------------

def computeFailures(successes, trials):
    N        = len(successes)
    failures = np.repeat(trials, N) - successes
    if any([ a<0 for a in failures]):
        raise ValueError("Number of trials is smaller than some counts.")
    return failures

def timingsToCounts(timings, binsize):
    MIN       = min(map(min, timings))
    MAX       = max(map(max, timings))
    N         = int(math.ceil(float(MAX-MIN)/binsize))
    successes = list(np.repeat(0, N+1))
    x         = range(MIN, MAX+binsize, binsize)
    for trial in timings:
        for t in trial:
            n = int(math.ceil(float(t-MIN)/binsize))
            successes[n] += 1
    trials    = len(timings)
    failures  = computeFailures(successes, trials)
    counts    = [successes, failures]
    return x, counts

def parseConfig(config_file):
    config_parser = ConfigParser.RawConfigParser()
    config_parser.read(config_file)

    if config_parser.sections() == []:
        raise IOError("Invalid configuration file.")
    if config_parser.has_section('Counts'):
        counts    = config.readMatrix(config_parser, 'Counts', 'counts', int)
        N         = len(counts[0])
        prior     = list(np.repeat(1, N))
        alpha     = config.readAlpha(config_parser, len(counts), N, 'Counts', int)
        prior     = config.readModelPrior(config_parser, N, 'Counts', int)
        options['script'] = config.readScript(config_parser, 'Counts', os.path.dirname(config_file))
        result    = bin(counts, alpha, prior)
        if options['save']:
            saveResult(result)
        else:
            vis.plotBinning(result, options)
            show()
    if config_parser.has_section('Trials'):
        binsize   = config_parser.getint('Trials', 'binsize')
        timings   = config.readMatrix(config_parser, 'Trials', 'timings', int)
        x, counts = timingsToCounts(timings, binsize)
        N         = len(counts[0])
        prior     = list(np.repeat(1, N))
        alpha     = config.readAlpha(config_parser, len(counts), N, 'Trials', int)
        prior     = config.readModelPrior(config_parser, N, 'Trials', int)
        options['script'] = config.readScript(config_parser, 'Trials', os.path.dirname(config_file))
        result    = bin(counts, alpha, prior)
        if options['save']:
            saveResult(result)
        else:
            vis.plotBinningSpikes(x, timings, result, options)
            show()

# main
# ------------------------------------------------------------------------------

options = {
    'epsilon'           : 0.00001,
    'marginal'          : 0,
    'marginal_step'     : 0.01,
    'marginal_range'    : (0.0,1.0),
    'n_moments'         : 3,
    'which'             : 0,
    'script'            : None,
    'load'              : None,
    'save'              : None,
    'verbose'           : False,
    'prombsTest'        : False,
    'compare'           : False,
    'bprob'             : False,
    'differential_gain' : False,
    'effective_counts'  : False,
    'multibin_entropy'  : False,
    'model_posterior'   : True,
    }

def main():
    global options
    try:
        longopts   = ["help", "verbose", "load=", "save=", "marginal", "marginal-range:"
                      "marginal-step=", "which=", "epsilon=", "moments=", "prombsTest",
                      "effective-counts"]
        opts, tail = getopt.getopt(sys.argv[1:], "demr:s:k:bhvt", longopts)
    except getopt.GetoptError:
        usage()
        return 2
    output = None
    for o, a in opts:
        if o in ("-v", "--verbose"):
            sys.stderr.write("Verbose mode turned on.\n")
            options["verbose"] = True
        if o in ("-t", "--prombsTest"):
            sys.stderr.write("Testing prombs.\n")
            options["prombsTest"] = True
        if o in ("-m", "--marginal"):
            options["marginal"] = True
        if o in ("-r", "--marginal-range"):
            options['marginal_range'] = tuple(map(float, a[1:-1].split(',')))
        if o in ("-s", "--marginal-step"):
            options["marginal_step"] = float(a)
        if o in ("-k", "--moments"):
            if int(a) >= 3:
                options["n_moments"] = int(a)
            else:
                usage()
                return 0
        if o in ("-h", "--help"):
            usage()
            return 0
        if o == "-b":
            options["bprob"] = True
        if o == "-d":
            options["differential_gain"] = True
        if o == "--effective-counts":
            options["effective_counts"] = True
        if o == "-e":
            options["multibin_entropy"] = True
        if o == "--load":
            options["load"] = a
        if o == "--save":
            options["save"] = a
        if o == "--which":
            options["which"] = int(a)
        if o == "--epsilon":
            options["epsilon"] = float(a)
    if len(tail) != 1:
        usage()
        return 1
    parseConfig(tail[0])
    return 0

if __name__ == "__main__":
    sys.exit(main())
